User/ryanbod/mock settings disk access

ryanbodrug-microsoft opened the pull request:
## Summary of the Pull Request

Dependency Inject SettingsUtils and SystemIOProvider to abstract disk access, and remove the need to write / read from disk during unit tests. 

## PR Checklist
* [x] Applies to #6096 
* [ ] CLA signed. If not, go over [here](https://cla.opensource.microsoft.com/microsoft/PowerToys) and sign the CLA
* [ ] Tests added/passed
* [ ] Requires documentation to be updated
* [ ] I've discussed this with core contributors already. If not checked, I'm ready to accept this work might be rejected in favor of a different grand plan. Issue number where discussion took place: #xxx

## Info on Pull Request

_What does this include?_
1. Introduces SystemIOProvider and IIOProvider interface to wrap methods that hit the disk. 
2. Dependency Inject the concrete implementation of SystemIOProvider and SettingsUtils into the viewmodels, so that unit tests can provide stub and mock implementations. 
3. Removes setup and cleanup methods from unit tests that write to disk.
4. Removed 'enabledbydefault' test for colour picker and added a new test that validates all modules are enabled by default. 

## Validation Steps Performed

_How does someone test & validate?_
1.  Runs all unit tests. 
2. Opens Settings for each module, 
 2.1 Verifies that settings update and save. 
 2.2 Verifies that modules are able to load modified settings.

ryanbodrug-microsoft committed:
1) Making Directory Methods private.
2) Removing the CreateDirectory / DeleteDirectory functionality from all Settings Unit Tests.

ryanbodrug-microsoft committed:
Abstracting disk access via IIOProvider to be able to provide mocks for unit tests instead of writing to disk.   This also prevents developers who are running unit tests from interfering with the PowerToys settings on their local dev box.

ryanbodrug-microsoft committed:
Dependency Injecting stub SettingsUtils for all tests

ryanbodrug-microsoft committed:
Removing ISettingsUtils from constructors of objects that need to be deserialized (ColorPickerSettings/PowerLauncherSettings) as this breaks System.Text.Json

ryanbodrug-microsoft committed:
Merge branch 'master' into user/ryanbod/mock_settings_disk_access

ryanbodrug-microsoft commented:
All previously observed issues.  Were not a result of these changes. 
1) Keyboard Manager process wasn't launching was fixed in. #6185. 
2) IPC messages that weren't getting picked up on settings change was caused by strings not being localized.  This just required a clean / rebuild.  See #6197 for context.

ryanbodrug-microsoft commented:
@laviusmotileng-ms:  ^^ FYI

arjunbalgovind commented on src/core/Microsoft.PowerToys.Settings.UI.Lib/Utilities/SystemIOProvider.cs:
```diff
@@ -0,0 +1,37 @@
+{
+    public class SystemIOProvider : IIOProvider
+    {
+        public bool CreateDirectory(string path)
```
[nit] I noticed the return value isn't used anywhere right now. Has it been added since it could be used in the future?

ryanbodrug-microsoft replied:
Yeah I put it in to give us the option to check. The current code base doesn't check so it will likely throw an exception on  the WriteAllText.  We can probably be more defensive about this (Potentially in alekhya's refactoring).  For now I was trying to minimize changes to logic, and we can improve the implementation seperately.

arjunbalgovind commented on src/core/Microsoft.PowerToys.Settings.UI.UnitTests/ModelsTests/BasePTModuleSettingsTest.cs:
```diff
@@ -3,9 +3,12 @@
 // See the LICENSE file in the project root for more information.
 
 using System;
+using System.Security.Cryptography.X509Certificates;
```
What is this used for?

ryanbodrug-microsoft replied:
It's not

arjunbalgovind commented on src/core/Microsoft.PowerToys.Settings.UI.UnitTests/ModelsTests/BasePTModuleSettingsTest.cs:
```diff
@@ -21,6 +24,21 @@ public class BasePTModuleSettingsTest
         [Obsolete]
         public void ToJsonString_ShouldReturnValidJSONOfModel_WhenSuccessful()
         {
+            //Mock Disk access
+            string saveContent = string.Empty;
+            string savePath = string.Empty;
+            var mockIOProvider = new Mock<IIOProvider>();
+            mockIOProvider.Setup(x => x.WriteAllText(It.IsAny<string>(), It.IsAny<string>()))
+                          .Callback<string, string>((path, content) =>
+                          {
+                              savePath = path;
+                              saveContent = content;
+                          });
+            mockIOProvider.Setup(x => x.ReadAllText(It.Is<string>(x => x.Equals(savePath, StringComparison.Ordinal))))
+                          .Returns(() => saveContent);
+
+            var settingsUtils = new SettingsUtils(mockIOProvider.Object);
+
```
Can `GetMockIOProviderForSaveLoadExists` be used here?

ryanbodrug-microsoft replied:
Yep good call 👍

arjunbalgovind commented on src/core/Microsoft.PowerToys.Settings.UI.UnitTests/Mocks/IIOProviderMocks.cs:
```diff
@@ -0,0 +1,38 @@
+            string savePath = string.Empty;
+            string saveContent = string.Empty;
+            var mockIOProvider = new Mock<IIOProvider>();
+            mockIOProvider.Setup(x => x.WriteAllText(It.IsAny<string>(), It.IsAny<string>()))
```
[Clarification] Will the savePath and saveContent variables be accessed by reference? I'm confused on why the mocked methods would work since savePath and saveContent would go out of reference at the end of this method and get garbage collected.

ryanbodrug-microsoft replied:
savePath and contentPath, being strings, are reference types, and therefore they will be marked by the garbage collector and won't be collected when the collection gets triggered.  Unlike C++ which would clean the memory up when it goes out of scope(unless it was RVOd), the memory used by savePath and contentPath won't get freed until memory pressure intiates garbage collection.

arjunbalgovind commented on src/core/Microsoft.PowerToys.Settings.UI.UnitTests/ViewModelTests/ImageResizer.cs:
```diff
@@ -226,5 +212,31 @@ public void UpdateWidthAndHeight_ShouldUpateSize_WhenCorrectValuesAreProvided()
             Assert.AreEqual(50, imageSize.Height);
         }
+
+        /// <summary>
+        /// This method mocks an IO provider to validate tests wich required saving to a file, and then reading the contents of that file, or verifying it exists
+        /// </summary>
+        /// <returns></returns>
+        Mock<IIOProvider> GetMockIOProviderForSaveLoadExists()
+        {
+            string savePath = string.Empty;
+            string saveContent = string.Empty;
+            var mockIOProvider = new Mock<IIOProvider>();
+            mockIOProvider.Setup(x => x.WriteAllText(It.IsAny<string>(), It.IsAny<string>()))
+                          .Callback<string, string>((path, content) =>
+                          {
+                              savePath = path;
+                              saveContent = content;
+                          });
+            mockIOProvider.Setup(x => x.ReadAllText(It.Is<string>(x => x.Equals(savePath, StringComparison.Ordinal))))
+                          .Returns(() => saveContent);
+
+            mockIOProvider.Setup(x => x.FileExists(It.Is<string>(x => x.Equals(savePath, StringComparison.Ordinal))))
+                          .Returns(true);
+            mockIOProvider.Setup(x => x.FileExists(It.Is<string>(x => !x.Equals(savePath, StringComparison.Ordinal))))
+                          .Returns(false);
+
+            return mockIOProvider;
+        }
```
Is this a redefinition?

ryanbodrug-microsoft replied:
Yep

arjunbalgovind reviewed (commented):
Validated that all the tests pass and settings get saved/loaded correctly on each page of the Settings app. Apart from the comments I've left, one part I'm confused about is the use of ISettingsUtils. It seems like this interface was required only for the mock done in `GetStubSettingsUtils`, but I feel like that functionality could be implemented using IIOProvider. The default json could be set in the saveContent and loaded from that. Are there any other areas which motivate the requirement on the two interfaces?

I can approve the PR after these clarifications.

ryanbodrug-microsoft committed:
Merging latest main and adding in DeleteSettings functionality to io provider and settings utils

ryanbodrug-microsoft committed:
Merge latest from main

ryanbodrug-microsoft committed:
Removing unused namespace reference

ryanbodrug-microsoft committed:
Removing redifined mock

ryanbodrug-microsoft committed:
As per PR feedback.  Stub Settings utils should work with any settings type if the intent is to compile / avoid null ref exceptions.

Strangely when implementing this fix it became apparent that a stub settings isn't enough, and disk access needed to be mocked.  I can't explain why the tests were passing previously.

ryanbodrug-microsoft commented on src/core/Microsoft.PowerToys.Settings.UI.UnitTests/Mocks/ISettingsUtilsMocks.cs:
```diff
@@ -0,0 +1,31 @@
+            settingsUtils.Setup(x => x.GetSettings<PowerRenameLocalProperties>(It.IsAny<string>(), It.IsAny<string>())).Returns(new PowerRenameLocalProperties());
+            settingsUtils.Setup(x => x.GetSettings<GeneralSettings>(It.IsAny<string>(), It.IsAny<string>())).Returns(new GeneralSettings());
+            settingsUtils.Setup(x => x.GetSettings<FancyZonesSettings>(It.IsAny<string>(), It.IsAny<string>())).Returns(new FancyZonesSettings());
+            settingsUtils.Setup(x => x.GetSettings<ShortcutGuideSettings>(It.IsAny<string>(), It.IsAny<string>())).Returns(new ShortcutGuideSettings());
```
Great callout.  I updated this method to leverage the It.IsAnyType. One interesting thing (I can't explain) is that for some reason the *Settings object that was previously getting returned seemed to be reused somehow.  When I modified it to use `Activator.CreateInstance(typeArgument);` instead of `new *Settings();` it exposed some tests that needed a mocked IO provider instead of a simple stub.   I've updated those tests as well.

ryanbodrug-microsoft committed:
Leveraging GetMockIOProviderForSaveLoadExists

ryanbodrug-microsoft commented:
> Validated that all the tests pass and settings get saved/loaded correctly on each page of the Settings app. Apart from the comments I've left, one part I'm confused about is the use of ISettingsUtils. It seems like this interface was required only for the mock done in `GetStubSettingsUtils`, but I feel like that functionality could be implemented using IIOProvider. The default json could be set in the saveContent and loaded from that. Are there any other areas which motivate the requirement on the two interfaces?
> 
> I can approve the PR after these clarifications.

A bit philosophical (shocking coming from me), but IIOProvider, and ISettingsUtils encapsulate different concerns to me.   The SettingsUtils should encapsulate the logic specific to saving and loading settings information, and any patterns that they may include, such as does the file exist (if not create it etc).    The IOProvider encapsulates the logic pertaining to persisting the information, hypothetically by writting to disk, or saving to a sql db, or blob storage.    This is intended to decouple the settings and unit tests from a specific implementation, but doesn't actually define the patterns that we should be using.  Alekhya's change should be doing that, and I may have to modify some of this once I take a look at her PR.

arjunbalgovind reviewed (approved):
LGTM!
